ibis/pandas/aggcontext.py

```diff
@@ -287,48 +292,71 @@ def compute_window_spec_default(obj, _):
 
 
 class Window(AggregationContext):
-
     __slots__ = 'construct_window',
 
     def __init__(self, kind, *args, **kwargs):
+        super(Window, self).__init__(
+            parent=kwargs.pop('parent', None),
+            group_by=kwargs.pop('group_by', None),
+            order_by=kwargs.pop('order_by', None),
+            dtype=kwargs.pop('dtype'),
+        )
         self.construct_window = operator.methodcaller(kind, *args, **kwargs)
 
     def agg(self, grouped_data, function, *args, **kwargs):
-        if callable(function):
-            return self.construct_window(grouped_data).apply(
-                _apply(function, args, kwargs)
-            )
-
-        if not isinstance(function, six.string_types):
-            raise TypeError(
-                '{} aggregation function must be a string or callable'.format(
-                    type(self).__name__
+        group_by = self.group_by
+
+        if not group_by:
+            windowed = self.construct_window(grouped_data)
+            if callable(function):
+                return windowed.apply(_apply(function, args, kwargs))
+            else:
+                assert isinstance(function, six.string_types)
+                method = getattr(windowed, function)
+                result = method(*args, **kwargs)
+                return result
+        else:
+            if callable(function):
+                method = functools.partial(
+                    operator.methodcaller('apply'),
+                    _apply(function, args, kwargs)
                 )
-            )
-
-        try:
-            method = self.short_circuit_method(grouped_data, function)
-        except AttributeError:
-            window = self.construct_window(grouped_data)
-            method = getattr(window, function)
-
-        result = method(*args, **kwargs)
+            else:
+                assert isinstance(function, six.string_types)
+                method = operator.methodcaller(function, *args, **kwargs)
+
+        order_by = self.order_by
+
+        keys = group_by + order_by
+        frame = self.parent.obj
+        name = grouped_data.obj.name
+        indexed_series = frame[keys + [name]].set_index(
```

Comment:
This block is pretty terse, comments? :)

Reply:
Yep